### PR TITLE
Implement catch-up phase

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
It's not ready for merging yet, but it can be looked at.

This still needs a bit of refactoring, separating atomic changes into their own commits.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
Also this only implements catch-up during booting up. The other definitions of done in the Catchup Phase Issue need to be solved.

File: source/agora/test/Ledger.d [hunk 1]
```diff
@@ -76,3 +77,47 @@ unittest
...
     assert(node_1.getBlocksFrom(1000, 10).length == 0);
 }
+
+/// test catch-up phase
+unittest
+{
+    import std.algorithm;
+    import std.conv;
+    import std.format;
+    import std.digest.sha;
```
Comment (Geod24) [inline source/agora/test/Ledger.d]: Should be removed

Comment (AndrejMitrovic) [reply source/agora/test/Ledger.d]: damn where did that sneak in from lol

**Comment by AndrejMitrovic** [comment]
I've atomic'ed (atomized?) the commits, it should look much nicer now.

I probably need to think of edge cases to add to the unittests.

File: source/agora/node/RemoteNode.d [hunk 2]
```diff
@@ -170,4 +170,25 @@ class RemoteNode
...
+    public ulong getBlockHeight ( )
+    {
+        try
+        {
+            return this.api.getBlockHeight();
+        }
+        catch (Exception ex)
+        {
+            logError(ex.message);
+            return 0;
```
Comment (Geod24) [inline source/agora/node/RemoteNode.d]: `ulong.max`

File: source/agora/test/Base.d [hunk 3]
```diff
@@ -183,10 +188,14 @@ public final class TestNode (Net) : Node!Net, TestAPI
...
     MemMetadata mem_metadata;
 
+    /// The ledger used for the tests
+    Ledger ledger;
+
     ///
     public this (Config config)
     {
         this.mem_metadata = new MemMetadata;
+        this.ledger = new Ledger;
```
Comment (Geod24) [inline source/agora/test/Base.d]: That code doesn't make sense to me. You instantiate a new `Ledger` just to override it with itself in the parent constructor. You also shadow the parent `ledger` variable ?

Comment (AndrejMitrovic) [reply source/agora/test/Base.d]: The shadowing was an oversight. I'll simplify this a bit tomorrow.

File: source/agora/node/RemoteNode.d (hunk 2 above)
Comment (AndrejMitrovic) [reply source/agora/node/RemoteNode.d]: Good point, block height is zero for genesis blocks. I'll add a test-case. Also like I said before I dislike this class, but I plan on making it better at a later point. There's no retry mechanism for example.

**Comment by AndrejMitrovic** [comment]
Update: Added periodic catch-up phase, and added a network test for it. This completes all of the tasks and should cover everything in the definition of done.

File: source/agora/test/Base.d [hunk 4]
```diff
@@ -175,6 +177,9 @@ public interface TestAPI : API
 
     ///
     public abstract void metaAddPeer(string peer);
+
+    ///
+    public abstract void addBlockToLedger (Block block);
```
Comment (Geod24) [inline source/agora/test/Base.d]: Why do you need this ? I don't see an override anywhere ?

File: source/agora/test/Ledger.d [hunk 5]
```diff
@@ -118,3 +118,48 @@ unittest
         assert(node.getBlockHeight() == 100);
     }
 }
+
+/// test catch-up phase after initial booting (periodic catch-up)
+unittest
+{
+    import core.thread;
+    import std.conv;
+    import vibe.core.log;
```
Comment (Geod24) [inline source/agora/test/Ledger.d]: Two leftover imports

File: source/agora/node/RemoteNode.d (hunk 2 above)
Comment (Geod24) [reply source/agora/node/RemoteNode.d]: Ping

**Review by Geod24 — CHANGES_REQUESTED** [review]
I would merge the last two commits. Also, do you want the ability to arbitrarily delay a node ? Because I can give you that power easily in `LocalRest`.

File: source/agora/test/Base.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/agora/test/Base.d]: Well, the question is how do you get access to the Ledger from the unittest. As it's required right now. edit: reworded

Comment (Geod24) [reply source/agora/test/Base.d]: You never access the internals of a node at the moment.

Comment (AndrejMitrovic) [reply source/agora/test/Base.d]: Well the first test is:

Instantiate the nodes. Don't start networking yet. One node (node A) needs to have 100 blocks, the 3 are empty.

Then, make them connect to each other. The 3 other nodes will retrieve blocks from the 1st node on boot.

But to make this work, I need to add those blocks to node A before it connects to the 3 other nodes.

**Comment by AndrejMitrovic** [comment]
I've squashed the two commits.

I'm thinking about how to get rid of those methods in the TestAPI now.

**Comment by AndrejMitrovic** [comment]
> Also, do you want the ability to arbitrarily delay a node ? Because I can give you that power easily in LocalRest.

Hmm.. that might be very useful.

**Comment by AndrejMitrovic** [comment]
All the Mac builds fail. Hmm..

**Comment by AndrejMitrovic** [comment]
It's a timing issue.

File: source/agora/test/Base.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/agora/test/Base.d]: To be more specific here. I can't just use the `sendTransaction` API here because this will trigger the gossip protocol to send the transaction over the whole network, and then each node will create their own block. So by the time the node tries to retrieve the blocks from the other nodes, it will already have all the blocks.

Delaying the node might not help. I would need some way to temporarily disable the gossip protocol, but then this becomes very messy to have all these special cases for unittesting.

The more features the node has, the more complicated it is to test a specific feature in the node without triggering an avalanche of other features.

**Comment by AndrejMitrovic** [comment]
Ok I made a more stable version of the tests that shouldn't suffer too much from timing issues. I added polling and a timeout of 4 seconds.

**Comment by AndrejMitrovic** [comment]
Very peculiar that macOS dmd 2.085.1 seems to consistently be failing now..

File: source/agora/test/Base.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/agora/test/Base.d]: I will be able to replace this code once LocalRest has the ability to filter requests. Then I can do something like:

- Use the sendTransaction API to one node
- Make the other nodes ignore transaction gossiping
- The other nodes will then catch-up using the periodic `getBlocksFrom` call.

This is way nicer than messing with Node's internals.

**Comment by AndrejMitrovic** [comment]
~@Geod24 I'm having consistent time-outs in the MacOS build: https://travis-ci.com/bpfkorea/agora/jobs/214998228

It seems it gets stuck trying to install DMD & Dub.~

And it's green.

**Comment by Geod24** [comment]
I guess you want to rebase ?

**Comment by AndrejMitrovic** [comment]
> I guess you want to rebase ?

Ah I thought this one was rebased.

**Comment by AndrejMitrovic** [comment]
I mean merged.

**Comment by AndrejMitrovic** [comment]
Rebased.

File: source/agora/node/Network.d [hunk 6]
```diff
@@ -137,6 +140,72 @@ public class NetworkManager
...
+
+    public void retrieveLatestBlocks (Ledger ledger)
+    {
+        this.runTask(
+        ()
+        {
+            // periodic task
+            while (1)
+            {
+                logInfo("Retrieving latest blocks..");
```
Comment (Geod24) [inline source/agora/node/Network.d]: I think this logging should only happen if we are actually catching up

File: source/agora/node/RemoteNode.d [hunk 7]
```diff
@@ -170,4 +171,55 @@ class RemoteNode
...
+
+    /***************************************************************************
+
+        Returns:
+            the height of the node's ledger,
+            or ulong.max if the request failed
+
+    ***************************************************************************/
+
+    public ulong getBlockHeight ( )
```
Comment (Geod24) [inline source/agora/node/RemoteNode.d]:
```suggestion
    public ulong getBlockHeight ()
```

File: source/agora/test/Ledger.d [hunk 8]
```diff
@@ -22,13 +23,24 @@ import agora.common.Hash;
 import agora.common.Transaction;
 import agora.test.Base;
 
+/// Returns: true if all node's block heights are equal to the provided height
+private bool allHeightsEqual ( TestAPI[] nodes, size_t height )
```
Comment (Geod24) [inline source/agora/test/Ledger.d]:
```suggestion
private bool allHeightsEqual (TestAPI[] nodes, size_t height)
```

**Review by Geod24 — CHANGES_REQUESTED** [review]
As discussed IRL: Bad rebase, some code is missing

**Comment by AndrejMitrovic** [comment]
Fixed the suggested changes.

**Comment by AndrejMitrovic** [comment]
Rebased to use new LocalRest sleep() feature.

However it's a bit ugly because now I have to store `RemoteAPI!TestAPI` in `TestNetworkManager`, not just `TestAPI`. And then there's the issue of `RemoteAPI!TestAPI[]` not implicitly converting to `TestAPI[]`.

**Comment by AndrejMitrovic** [comment]
Edit: Pushed a simplification https://github.com/bpfkorea/agora/compare/3c38af6c675d18cab84aa8c89c702d6b3cf8bb7c..bf03c16e9bc1267934450648f55a9f583ffbc9e1

File: source/agora/node/Network.d [hunk 9]
```diff
@@ -137,6 +140,72 @@ public class NetworkManager
...
+        Retrieve blocks starting from block_height up to the highest block
+        that's available from the connected nodes.
+
+        Params:
+            block_height = the starting block height to begin retrieval from
+
+    ***************************************************************************/
+
+    private void getBlocksFrom (ulong block_height,
+        void delegate(Block[]) @safe onReceivedBlocks)
```
Comment (Geod24) [inline source/agora/node/Network.d]:
```suggestion
        scope void delegate(Block[]) @safe onReceivedBlocks)
```

File: source/agora/node/Node.d [hunk 10]
```diff
@@ -174,4 +176,18 @@ public class Node (Network) : API
...
+        Returns a new Ledger instance. Can be overriden in unittests.
+
+        Returns:
+            a Ledger instance
+
+    ***************************************************************************/
+
+    protected Ledger getLedger ()
+    {
+        return new Ledger();
```
Comment (Geod24) [inline source/agora/node/Node.d]: I don't think you need this anymore ?

File: source/agora/test/Ledger.d [hunk 11]
```diff
@@ -71,3 +83,79 @@ unittest
...
+        Thread.sleep(100.msecs);
+    }
+}
+
+/// test catch-up phase after initial booting (periodic catch-up)
+unittest
+{
+    import core.thread;
+    import std.algorithm;
+    import std.format;
```
Comment (Geod24) [inline source/agora/test/Ledger.d]: Those 2 imports are not necessary anymore

File: source/agora/test/Ledger.d [hunk 12]
```diff
@@ -22,13 +23,24 @@ import agora.common.Hash;
 import agora.common.Transaction;
 import agora.test.Base;
 
+/// Returns: true if all node's block heights are equal to the provided height
+private bool allHeightsEqual (API)(API[] nodes, size_t height)
```
Comment (Geod24) [inline source/agora/test/Ledger.d]: Suggestion: `isConsistent` and check the block headers for equality ?

File: source/agora/test/Ledger.d (hunk 11 above)
Comment (AndrejMitrovic) [reply source/agora/test/Ledger.d]: I need `each` still.

**Comment by AndrejMitrovic** [comment]
Updated with suggested changes.

File: source/agora/test/Ledger.d (hunk 12 above)
Comment (AndrejMitrovic) [reply source/agora/test/Ledger.d]: Ok.

File: source/agora/test/Ledger.d [hunk 13]
```diff
@@ -71,3 +83,78 @@ unittest
...
     assert(node_1.getBlocksFrom(1000, 10).length == 0);
 }
+
+/// test catch-up phase during booting
+unittest
+{
+    import core.thread;
+    import std.algorithm;
+    import std.conv;
+    import std.format;
```
Comment (Geod24) [inline source/agora/test/Ledger.d]: conv and format superfluous

**Review by Geod24 — APPROVED** [review]
LGTM

**Comment by AndrejMitrovic** [comment]
Pushed a small update.

**Review by Geod24 — APPROVED** [review]
